cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -349,6 +370,47 @@ public Void call() throws Exception {
     });
   }
 
+  private ProgramState getProgramState(boolean success, String failureInfo) {
+    if (stopRequested) {
+      // Program explicitly stopped, return KILLED state
+      return new ProgramState(ProgramStatus.KILLED, null);
+    }
+    if (!success) {
+      // Program is unsuccessful, return FAILED state
+      return new ProgramState(ProgramStatus.FAILED, failureInfo);
+    }
+    // Program is successfully completed, return COMPLETE state
+    return new ProgramState(ProgramStatus.COMPLETED, null);
+  }
+
+  /**
+   * Calls the destroy method of {@link ProgramLifecycle}.
+   */
+  private void destroy(final boolean succeeded, final BasicSparkClientContext context,
```

Comment:
Seems like this method is quite similar to the onFinish. Can you use the same method and based on the type to decide which one to call?